backend/app/tests/routers/helpers/test_helpers.py

```diff
@@ -0,0 +1,139 @@
+from fastapi import FastAPI, Request  
+from contextlib import asynccontextmanager 
+from fastapi.testclient import TestClient
+
+from app.routers.helpers.helpers import * 
+from app.storage.storage import storage as wikipedia_storage
+from app.anti_recommenders.anti_recommender_proxy import AntiRecommenderProxy
+
+@asynccontextmanager 
+async def lifespan(app: FastAPI):
+    """
+    A lifespan event to test the storage of wikipedia data in the app state.
+    """
+
+    app.state.wikipedia_storage =   {
+        "info": "test-wikipedia-article"
+        }
+
+    yield 
+
+    app.state.wikipedia_storage = None 
+
+# intialize the FastAPI app and pass in the lifespan event.
+app = FastAPI(lifespan = lifespan)
+
+@app.get("/")
+async def get_app_state(request: Request):
+    """
+    A path operation function of the root endpoint.
+    It returns the contents of the app state via the access_app_state method.  
+    """
+    
+    return access_app_state(request)
+
+
+def test_access_app_state() -> None:
+    """
+    Test to check that data is correctly stored in the app state and can be accessed during runtime.
+    """
+    
+    with TestClient(app) as client:
+        app_state_dictionary = client.get("/")
+        assert app_state_dictionary.status_code == 200
+
+        # Check that the key "wikipedia_storage is correctly stored in the app state
+        assert app_state_dictionary.json()[0]["wikipedia_storage"] == {"info": "test-wikipedia-article"}
+    
+
+def test_generate_anti_recommendations_of_wikipedia_articles(mocker):
+    """
+    Test to check that the generate_wikipedia_articles method returns 
+    a tuple of records if the anti-recommendations of an article are found in storage.
+    """
+
+    # mock variables
+    mock_request = mocker.Mock()
+
+    # variables to mock return values
+    mock_article_records = tuple([{"article-record-1" :  {"info": "wikipedia-article-1"}}])
+    mock_anti_recommender_proxy = AntiRecommenderProxy()
+    
+    # Mock a function
+    mocker.patch("app.routers.helpers.helpers.access_app_state",
+                                        return_value = tuple([{"wikipedia_storage": wikipedia_storage}]))
+
+    # Mock an object and its methods
+    mocker.patch.object(AntiRecommenderProxy, "get_title")
```

Comment:
Having to mock so many methods suggests that the underlying logic that `generate_wikipedia_articles` uses needs to be simplified.

Can you think of this in a more functional way than directly the storage? Similar to Elm?

Ideally, the generate function (1) receives the current (per-user) state from "somewhere" and (2) emits/returns articles as well as a new state.

It's not pure because it needs to reach out to the AntiRecommender (which has side effects), but then you can just mock the AntiRecommender.

Reply:
Yes, this was helpful. Thank you. I have improved the logic in `StorageManager` and `Storage`. I am not entirely sure if they are in their simplest forms, but I have reduced the number of mock methods in test files.